Add DTS partition support

agross-oss opened the pull request:
This set of patches implements the DTS partitions support for defining flash partitions.

Commit:
dts/extract_dts_includes.py: Add partition support

This patch adds parsing support for flash partitions.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

d3zd3z commented on scripts/extract_dts_includes.py:
```diff
@@ -240,9 +240,12 @@ def extract_reg_prop(node_address, names, defs, def_label, div):
     address_cells = reduced[address]['props'].get('#address-cells', address_cells)
     size_cells = reduced[address]['props'].get('#size-cells', size_cells)
 
+  if post_label == None:
```
Either:
  if post_label is None:
or
  if not post_label:

d3zd3z commented on scripts/extract_dts_includes.py:
```diff
@@ -397,14 +400,21 @@ def extract_single(node_address, yaml, prop, key, prefix, defs, def_label):
-  def_label = convert_string_to_label(get_compat(node)).upper()
-  def_label += '_' + node_address.split('@')[-1].upper()
+
+  if yaml[node_compat].get('base_label') != None:
```
... is not None

Commit:
kconfig: misc: Conditionally remove FLASH_LOAD options

This patch adds a condition on the FLASH_LOAD_OFFSET and FLASH_LOAD_SIZE
to make them conditionally excluded if HAS_DTS is enabled.  FLASH
partition information must be defined for all DTS platforms which
utilize flash.

For DTS enabled platforms, CONFIG_FLASH_LOAD_OFFSET and
CONFIG_FLASH_LOAD_SIZE will be generated from the flash information
defined in the DTS file.  The values used for these variables will be
determined by the zephyr,code-partition chosen node.  If no chosen
node is specified, the zephyr,flash chosen node will be used.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

d3zd3z reviewed (approved):
This looks good to me, I've also tested this works with mcuboot and applications, allowing the partitioning information to be removed from the bootloader itself.

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -397,14 +400,21 @@ def extract_single(node_address, yaml, prop, key, prefix, defs, def_label):
-  def_label = convert_string_to_label(get_compat(node)).upper()
-  def_label += '_' + node_address.split('@')[-1].upper()
+
+  if yaml[node_compat].get('base_label') is not None:
```
Would `if 'base_label' in yaml[node_compat]` work? Might be a bit more idiomatic.

agross-linaro replied:
yeah that is clearer

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -397,14 +400,21 @@ def extract_single(node_address, yaml, prop, key, prefix, defs, def_label):
+    def_label = convert_string_to_label(node_compat.upper())
+    def_label += '_' + node_address.split('@')[-1].upper()
+
+  if label_override != None:
```
PEP008: "Comparisons to singletons like None should always be done with is or is not , never the equality operators."

https://www.python.org/dev/peps/pep-0008/

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
-def extract_node_include_info(reduced, node_address, yaml, defs, structs):
-  node = reduced[node_address]
-  node_compat = get_compat(node)
+def extract_node_include_info(reduced, root_node_address, sub_node_address, yaml, defs, structs, y_sub):
```
PEP 008: "Limit all lines to a maximum of 79 characters."

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
     return {}, {}
 
-  y_node = yaml[node_compat]
+  if y_sub == None:
```
is None

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
+  else:
+    y_node = y_sub
+
+  if yaml[node_compat].get('use-property-label', False) != False:
```
PEP008: "Don't compare boolean values to True or False using == ."

(I.e., `if if yaml[node_compat].get('use-property-label', False):`.)

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
+
+  if yaml[node_compat].get('use-property-label', False) != False:
+    for yp in y_node['properties']:
+      if yp.get('label') != None:
```
is not None

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
+  if yaml[node_compat].get('use-property-label', False) != False:
+    for yp in y_node['properties']:
+      if yp.get('label') != None:
+        if node['props'].get('label') != None:
```
is not None

mbolivar commented on scripts/extract_dts_includes.py:
```diff
@@ -415,24 +425,39 @@ def extract_property(yaml, node_address, y_key, y_val, names, prefix, defs):
   for yp in y_node['properties']:
     for k,v in yp.items():
+      if 'properties' in v:
+        for c in reduced.keys():
```
`.keys()` is not necessary; dictionary iterators are over keys by default.

```python
In [1]: d = {1:2, 3:4}

In [2]: for k in d:
   ...:     print(k)
   ...:     
1
3
```

mbolivar commented on misc/Kconfig:
```diff
@@ -15,6 +15,7 @@ config HAS_FLASH_LOAD_OFFSET
 	This option is selected by targets having a FLASH_LOAD_OFFSET
 	and FLASH_LOAD_SIZE.
 
+if !HAS_DTS
```
Will this break existing targets which have DTS and want to use flash, but don't define partitions defined yet? If so, could merging this patch be delayed until we've had time to fix those up?

agross-linaro replied:
no it won't break those.  It defaults to using flash0 information.

mbolivar replied:
OK, thanks.

mbolivar reviewed (changes requested):
Some Python code style nitpicks and a question about existing applications on the last patch.

By the way, you can run a PEP008 linter on Python code, similarly to checkpatch:

https://pypi.python.org/pypi/pep8

d3zd3z commented:
While we're on PEP-8, we should probably fix the indentation to be 4 spaces. But, I would argue that should be a separate patch.

agross-linaro commented:
yeah i was going to do a pep-8 blanket patch modification separately.

agross-linaro commented:
last commit fixed all the PEP-8 comments from Marti

mbolivar reviewed (commented):
Thanks for touching those up. Looking forward to having partition support merged -- we need it to build multiple applications that can be chainloaded by mcuboot without resorting to hacks and duplicated code.

galak commented on dts/arm/yaml/partition.yaml:
```diff
@@ -0,0 +1,43 @@
+      properties:
+        - reg:
+            type: array
+            description: mmio register space
```
these seems wrong, while is mmio space, saying its register space seems a little misleading. (I assume we are talking about flashes that are MMIO direct access)

galak reviewed (changes requested):
Can you add to commit message details about how we are setting CONFIG_FLASH_LOAD.

Also, not sure best place, but would be nice to see example or doc for what partition info looks like in dts.

d3zd3z commented:
I have a partition map for k64f I've been using, should we include that somewhere in the docs? Also, fairly soon after this goes in, we'll be adding partitions to many of the devices.

agross-linaro commented:
Addressed comments and fixed one issue with making the partitions.yaml file common.

Commit:
dts: common: yaml: Add partition file

This patch adds the yaml definition for flash partitions.

Signed-off-by: Andy Gross <andy.gross@linaro.org>

galak commented:
@d3zd3z can you provide a pointer to the k64 partition.  Would like to see as an example.
